### DISCUSSION TITLE
Unable to use RPC within PHP

### DISCUSSION THREAD

**Discussion by awhitetiger** [transmission/transmission#4082] [Support]
### What is the issue?

I am unable to use my server's RPC when sending a post request in PHP. I only get "Bad Request" back. I have duplicated this request in both python and javascript and both work in those respective languages.
My PHP request: 
![image](https://user-images.githubusercontent.com/9219901/199516619-5075fceb-cabe-4831-912f-5b5e535abb19.png)

Response: 
![image](https://user-images.githubusercontent.com/9219901/199516739-0880d1d9-7fc1-4912-b2f4-046dd31326d5.png)

### Which application of Transmission?

transmission-remote

### Which version of Transmission?

3.0

**Comment by mikedld** [comment]
Works for me with PHP 8.1.11 and recent git, prints
```
{"arguments":{"version":"4.0.0-beta.1.dev (49a3813a23)"},"result":"success"}
```

If you're saying the same logic implemented in Python and JavaScript works then I suppose your configuration allows for requests to 192.168.1.109:9093, but maybe double-check the IP whitelisting to include the server handling the request, not the client opening the web page. You also didn't provide the implementation of `getTransmissionHeader()`, hopefully it's correct.

**Reply by awhitetiger** [reply]
Interesting enough, that function returns a 409 (as expected) and correctly passes the session-id as expected. My implementation of that:
![image](https://user-images.githubusercontent.com/9219901/199532215-59755532-f7a0-4b38-8353-8383bb7258cb.png)

Is there any log in transmission where it would spit out the reason its sending back a 400?

As far as the client handling the request, that may be it actually...its running in a docker on technically the same ip address, same server as transmission. Would I have to use localhost maybe? or perhaps is there maybe some conflict of transmission being at 192.168.1.109:9093 and my web server being at 192.168.1.109:8005 ?

**Comment by mikedld** [comment]
400 is not one of the responses Transmission would send, judging by calls to `evhttp_send_reply()` (directly) and `rpc_response_func()` (indirectly) in https://github.com/transmission/transmission/blob/3.00/libtransmission/rpc-server.c...

**Reply by awhitetiger** [reply]
I found the issue and it is my fault(obviously). For some reason curl does not like having a dynamic header value.
`array('Content-Type:application/json','Accept:application/jsonrequest','X-Transmission-Session-Id: '.$tokenID,'charset:UTF-8');` <--- returns a 400
`array('Content-Type:application/json','Accept:application/jsonrequest','X-Transmission-Session-Id:XXXX','charset:UTF-8');` <---- works

So for some reason appending that variable of the session ID borks it. good to know.. Thanks for helping me rack my brain on this. Edit: seems my variable had some endline characters :)) rtrim fixed it.
